README.md
# 20-40k_xbu


pw-bulk-edit/changelog.txt
= 2.92 =
* Confirmed compatibility with WooCommerce 6.0

= 2.91 =
* Confirmed compatibility with WooCommerce v5.9

= 2.90 =
* Added a new wp-config.php option to change the memory limit setting or disable it entirely: PWBE_MEMORY_LIMIT

= 2.89 =
* Confirmed compatibility with WordPress v5.8 and WooCommerce v5.5

= 2.88 =
* Fixed an issue where the check/uncheck all Checkbox would not work when scrolled down.

= 2.87 =
* Fixed an issue preventing string filters that contain an underscore character.

= 2.86 =
* Confirmed compatibility with WordPress 5.7 and WooCommerce 5.1

= 2.85 =
* Updated call to the WooCommerce hook woocommerce_update_product_variation to include the Product variable. Fixed an issue when bulk editing certain fields such as Product Description. Improved HTML to make it more valid. Updated readme.txt to reflect new features in the Pro version.

= 2.84 =
* Confirmed WooCommerce 5.0 compatibility.

= 2.83 =
* Confirmed compatibility with WooCommerce v4.9

= 2.82 =
* Updated call to woocommerce_update_product for other plugins that listen to this hook and use the new second parameter.

= 2.81 =
* Confirmed compatibility with WordPress 5.6 and WooCommerce 4.8

= 2.80 =
* Removed Javascript warning about missing source map files when the developer console is open. Confirmed compatibility with WooCommerce v4.7

= 2.79 =
* Confirmed compatibility with WooCommerce v4.6

= 2.78 =
* Verified compatibility with WooCommerce 4.5

= 2.77 =
* Fixed a potential error about a duplicate primary key when filtering.

= 2.76 =
* Fixed an issue that can cause the error 'pwbe_products table does not exist' on certain hosts, in particular sites hosted on wordpress.com

= 2.75 =
* Confirmed compatibility with WooCommerce v4.3

= 2.74 =
* When sorting numeric columns, use DECIMAL instead of SIGNED to make sorting fractions work properly.

= 2.73 =
* Small fix to not add an undo/redo entry if a bulk edit operation does not change any fields. Confirmed compatibility with WordPress 5.5

= 2.72 =
* Fixed an HTML tag with the Cancel button.

= 2.71 =
* Allow changing the required capability for the bulk editor by setting PWBE_REQUIRES_CAPABILITY constant in your wp-config.php. The default is manage_woocommerce.

= 2.70 =
* Fixed sorting all Number field types for databases that have the data as a string rather than an integer.

= 2.69 =
* Fixed sorting by Menu Order for databases that have the menu data as a string rather than an integer.

= 2.68 =
* Confirmed compatibility with WooCommerce v4.1

= 2.67 =
* Confirmed compatibility with WordPress 5.4

= 2.66 =
* Allow HTML in fields such as Description, Name, etc.

= 2.65 =
* SECURITY ALERT - This version fixes a reported security issue with the plugin. It is recommended that all users upgrade to this latest version.

= 2.64 =
* Confirmed compatibility with WordPress 5.3.2 and WooCommerce 4.0.0

= 2.63 =
* Confirmed compatibility with WooCommerce 3.9.0

= 2.62 =
* New feature: Copy a view. Integration with LiteSpeed Cache to purge the cache after saving products.

= 2.61 =
* Confirmed compatibility with the upcoming WooCommerce v3.8.0 release. Added a new wp-config option: PWBE_PREFILTER_VARIATIONS. If the data incorrectly contains product_variation records that are children of Simple Products you will want to enable this flag. It is disabled by default because typically it will not be needed and could slow down the search results.

= 2.60 =
* Added the ability to clear Number values (set to n/a) from the bulk edit menu.

= 2.59 =
* Fixed sorting by Catalog Visibility and plugin version mismatch.

= 2.58 =
* No release, plugin version mismatch.

= 2.57 =
* No release, plugin version mismatch.

= 2.56 =
* Confirmed compatibility with WooCommerce v3.6.4 and WordPress v5.2.1

= 2.55 =
* Made Variation Description part of the Standard Columns to avoid confusion.

= 2.54 =
* Updated currency bulk editor to support a higher decimal precision.

= 2.53 =
* Fixed a potential unitialized string offset error while saving products.

= 2.52 =
* Confirmed compatibility with WooCommerce v3.5.0

= 2.51 =
* Sync the Stock Status when changing stock quantity.

= 2.50 =
* Fixed a bug introduced in v2.49 that prevented some of the filters from being used.

= 2.49 =
* Added translation support using Loco Translate, Poedit, etc.

= 2.48 =
* Added new config option PWBE_SAVE_BATCH_SIZE to help prevent HTTP 504 Timeout errors with slow servers.

= 2.47 =
* Updated plugin to be able to handle malformed prices when doing bulk operations.

= 2.46 =
* Variations titles now show the formatted name.

= 2.45 =
* Fixed possible issue sorting on variation attribute.

= 2.44 =
* Fixed issue where number fields may appear blank for editing instead of "n/a".

= 2.43 =
* Tweak to SQL query when saving products to prevent potential error on some systems.

= 2.42 =
* Tweaks to fix blank MySQL error.

= 2.41 =
* Moved the plugin init code out of the woocommerce_init hook and back into the constructor.

= 2.40 =
* Added .pot file for translation support.

= 2.39 =
* Added PW Bulk Edit menu under the WooCommerce Products menu to make it easier to find.

= 2.38 =
* Fixed issue where Stock Status column wouldn't appear unless you had Enable Stock Management enabled.

= 2.37 =
* Fixed sorting on the Featured column in WooCommerce 3.0 and later.

= 2.36 =
* Fixed issue setting Featured flag and Catalog Visibility in WooCommerce 3.0 and later.

= 2.35 =
* Fixed issue with the Featured flag. Fixed a possible exception when filtering results.

= 2.34 =
* Removed a warning about an undefined variable.

= 2.33 =
* Updated the admin menu icon style.

= 2.32 =
* Updated the admin menu icon.

= 2.31 =
* Display 0 in the results for prices that are not blank.

= 2.30 =

pw-bulk-edit/pw-bulk-edit.php
<?php
/**
 * Plugin Name: PW WooCommerce Bulk Edit
 * Plugin URI: https://www.pimwick.com/pw-bulk-edit/
 * Description: A powerful way to update your WooCommerce product catalog. Finally, no more tedious clicking through countless pages making the same change to all products!
 * Version: 2.92
 * Author: Pimwick, LLC
 * Author URI: https://www.pimwick.com
 * Text Domain: pw-bulk-edit
 * Domain Path: /languages
 *
 * WC requires at least: 2.6.5
 * WC tested up to: 6.0
 *
*/
define('PWBE_VERSION', '2.92');

/*
Copyright (C) 2016-2018 Pimwick, LLC

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

// Exit if accessed directly
defined( 'ABSPATH' ) or exit;

// Increase the available memory since returning lots of data can often exhaust typical memory allocation amounts.
defined( 'PWBE_MEMORY_LIMIT' ) or define( 'PWBE_MEMORY_LIMIT', '1024M' );
if ( PWBE_MEMORY_LIMIT !== false ) {
    ini_set( 'memory_limit', PWBE_MEMORY_LIMIT );
    defined( 'WP_MEMORY_LIMIT' ) or define( 'WP_MEMORY_LIMIT', PWBE_MEMORY_LIMIT );
}

// Only change this if you are comfortable with possible unexpected behavior!
defined( 'PWBE_MAX_RESULTS' ) or define( 'PWBE_MAX_RESULTS', '1000' );

// Number of fields to save in a single AJAX call. Lower number can avoid HTTP 504 Timeout errors.
defined( 'PWBE_SAVE_BATCH_SIZE' ) or define( 'PWBE_SAVE_BATCH_SIZE', 25 );

// If the data contains product_variation records that are children of Simple Products you will want to
// enable this flag. This could make the query run slower.
defined( 'PWBE_PREFILTER_VARIATIONS' ) or define( 'PWBE_PREFILTER_VARIATIONS', false );

defined( 'PWBE_REQUIRES_CAPABILITY' ) or define( 'PWBE_REQUIRES_CAPABILITY', 'manage_woocommerce' );

// Verify this isn't called directly.
if ( !function_exists( 'add_action' ) ) {
    echo 'Hi there!  I\'m just a plugin, not much I can do when called directly.';
    exit;
}

if ( ! class_exists( 'PW_Bulk_Edit' ) ) :

register_uninstall_hook( __FILE__, array( 'PW_Bulk_Edit', 'plugin_uninstall' ) );

final class PW_Bulk_Edit {

    const NULL = '!!pwbe_null_value!!';

    static $options = array(
        'pwbe_help_dismiss_intro',
        'pwbe_help_minimize_filter_help',
        'pwbe_views',
        'pwbe_selected_view'
    );

    function __construct() {
        add_action( 'plugins_loaded', array( $this, 'plugins_loaded' ) );
        add_action( 'woocommerce_init', array( $this, 'woocommerce_init' ) );
    }

    function plugins_loaded() {
        load_plugin_textdomain( 'pw-bulk-edit', false, basename( dirname( __FILE__ ) ) . '/languages' );
    }

    function woocommerce_init() {
        remove_action( 'admin_print_scripts', 'print_emoji_detection_script' );
        remove_action( 'admin_print_styles', 'print_emoji_styles' );

        if ( is_admin() && current_user_can( get_option( 'pwbe_requires_capability', PWBE_REQUIRES_CAPABILITY ) ) ) {
            require( 'includes/columns.php' );
            require( 'includes/db.php' );
            require( 'includes/filters.php' );
            require( 'includes/select-options.php' );
            require( 'includes/sql-builder.php' );
            require( 'includes/views.php' );

            add_action( 'admin_menu', array( $this, 'register_admin_menu' ) );
            add_action( 'admin_enqueue_scripts', array( $this, 'admin_scripts' ), 9999 );
            add_action( 'wp_ajax_pwbe_options', array( $this, 'ajax_options' ) );
            add_action( 'wp_ajax_pwbe_filter_results', array( $this, 'ajax_filter_results' ) );
            add_action( 'wp_ajax_pwbe_get_view', array( $this, 'ajax_get_view' ) );
            add_action( 'wp_ajax_pwbe_save_view', array( $this, 'ajax_save_view' ) );
            add_action( 'wp_ajax_pwbe_delete_view', array( $this, 'ajax_delete_view' ) );
            add_action( 'wp_ajax_pwbe_save_products', array( $this, 'ajax_save_products' ) );
            add_action( 'wp_ajax_pwbe_get_save_products_error', array( $this, 'ajax_get_save_products_error' ) );
        }
    }

    public static function wc_min_version( $version ) {
        return version_compare( WC()->version, $version, ">=" );
    }

    public static function plugin_uninstall() {
        if ( ! current_user_can( 'activate_plugins' ) )
            return;

        foreach( PW_Bulk_Edit::$options as $option ) {
            delete_option( $option );
        }
    }

    function index() {
        $data = get_plugin_data( __FILE__ );
        $version = $data['Version'];
        $help_url = plugins_url( '/docs/index.html', __FILE__ );

        require( 'ui/index.php' );
    }

    function ajax_options() {
        $name = $_POST['option_name'];
        $value = $_POST['option_value'];

        if ( in_array( $name, PW_Bulk_Edit::$options ) ) {
            update_option( $name, $value );
        }
    }

    function ajax_filter_results() {
        require( 'ui/results.php' );
        wp_die();
    }

    function ajax_get_view() {
        $view_name = stripslashes( $_POST['name'] );

        update_option( 'pwbe_selected_view', $view_name );

        $views = PWBE_Views::get();
        $view = $views[ $view_name ];

        wp_send_json( $view );

        wp_die();
    }

    function ajax_save_view() {
        $option_value = get_option( 'pwbe_views' );
        $views = maybe_unserialize( $option_value );

        $clean_name = stripslashes( $_POST['name'] );

        $views[ $clean_name ] = stripslashes( $_POST['view_data'] );

        ksort( $views, SORT_NATURAL );

        update_option( 'pwbe_views', $views );
        update_option( 'pwbe_selected_view', $clean_name );

        wp_die();
    }

    function ajax_delete_view() {
        $option_value = get_option( 'pwbe_views' );
        $views = maybe_unserialize( $option_value );

        $view_name = stripslashes( $_POST['name'] );

        unset( $views[ $view_name ] );

        update_option( 'pwbe_views', $views );
        update_option( 'pwbe_selected_view', '' );

        wp_die();
    }

    function ajax_save_products() {
        register_shutdown_function( array( $this, 'save_products_exception' ) );

        check_ajax_referer( 'pw-bulk-edit-save-products', 'security' );

        require( 'includes/save-products.php' );

        if ( isset( $_POST['fields'] ) ) {
            $fields = $_POST['fields'];

            $save = new PWBE_Save_Products();
            $results = $save->save( $fields );

            require( 'ui/products-saved.php' );
        }

        wp_die();
    }

    function ajax_get_save_products_error() {
        $error_file = plugin_dir_path( __FILE__ ) . 'logs/save_products_exception.txt';

        printf( __( 'Error while saving products: %', 'pw-bulk-edit' ), file_get_contents( $error_file ) );

        wp_die();
    }

    public function save_products_exception() {
        $errfile = 'unknown file';
        $errstr  = 'shutdown';
        $errno   = E_CORE_ERROR;
        $errline = 0;

        $error = error_get_last();

        if ( $error !== NULL ) {
            $errno   = $error['type'];
            $errfile = $error['file'];
            $errline = $error['line'];
            $errstr  = $error['message'];

            if ( PW_Bulk_Edit::starts_with( plugin_dir_path( __FILE__ ), $errfile ) ) {
                $output_dir = plugin_dir_path( __FILE__ ) . 'logs';
                if ( ! file_exists( $output_dir ) ) {
                    mkdir( $output_dir, 0777, true );
                }
                file_put_contents( $output_dir . '/save_products_exception.txt', "$errstr in $errfile on line $errline" );
            }
        }
    }

    function error( $message ) {
        ?>
        <div class="error">
            <p><?php _e( $message, 'pw-bulk-edit' ); ?></p>
        </div>
        <?php
    }

    function register_admin_menu() {
        if ( empty ( $GLOBALS['admin_page_hooks']['pimwick'] ) ) {
            add_menu_page(
                __( 'PW Bulk Edit', 'pw-bulk-edit' ),
                __( 'Pimwick Plugins', 'pw-bulk-edit' ),
                get_option( 'pwbe_requires_capability', PWBE_REQUIRES_CAPABILITY ),
                'pimwick',
                array( $this, 'index' ),
                plugins_url( '/assets/images/pimwick-icon-120x120.png', __FILE__ ),
                6
            );

            add_submenu_page(
                'pimwick',
                __( 'PW Bulk Edit', 'pw-bulk-edit' ),
                __( 'Pimwick Plugins', 'pw-bulk-edit' ),
                get_option( 'pwbe_requires_capability', PWBE_REQUIRES_CAPABILITY ),
                'pimwick',
                array( $this, 'index' )
            );

            remove_submenu_page('pimwick','pimwick');
        }

        add_submenu_page(
            'pimwick',
            __( 'PW Bulk Edit', 'pw-bulk-edit' ),
            __( 'PW Bulk Edit', 'pw-bulk-edit' ),
            get_option( 'pwbe_requires_capability', PWBE_REQUIRES_CAPABILITY ),
            'pw-bulk-edit',
            array( $this, 'index' )
        );

        add_submenu_page(
            'edit.php?post_type=product',
            __( 'PW Bulk Edit', 'pw-bulk-edit' ),
            __( 'PW Bulk Edit', 'pw-bulk-edit' ),
            get_option( 'pwbe_requires_capability', PWBE_REQUIRES_CAPABILITY ),
            'wc-pw-bulk-edit',
            array( $this, 'index' )
        );
    }

    function admin_scripts( $hook ) {
        if ( !empty( $hook ) && substr( $hook, -strlen( 'pw-bulk-edit' ) ) === 'pw-bulk-edit' ) {
            wp_register_style( 'pwbe-font-awesome', plugins_url( '/assets/css/font-awesome.min.css', __FILE__ ), array(), PWBE_VERSION ); // 4.6.3
            wp_enqueue_style( 'pwbe-font-awesome' );

            wp_register_style( 'pwbe-select2', plugins_url( '/assets/css/select2.min.css', __FILE__ ), array(), PWBE_VERSION ); // 4.0.3
            wp_enqueue_style( 'pwbe-select2' );

            wp_register_style( 'pwbe-context-menu', plugins_url( '/assets/css/jquery.contextMenu.min.css', __FILE__ ), array(), PWBE_VERSION );
            wp_enqueue_style( 'pwbe-context-menu' );

            wp_enqueue_script( 'pwbe-select2', plugins_url( '/assets/js/select2.min.js', __FILE__ ), array(), PWBE_VERSION ); // 4.0.3

            wp_register_style( 'pw-bulk-edit', plugins_url( '/assets/css/style.css', __FILE__ ), array(), PWBE_VERSION );
            wp_enqueue_style( 'pw-bulk-edit' );

            wp_enqueue_script( 'pwbe-context-menu', plugins_url( '/assets/js/jquery.contextMenu.min.js', __FILE__ ), array( 'jquery-ui-position' ), PWBE_VERSION ); // 2.2.3

            wp_enqueue_script( 'pwbe-filters', plugins_url( '/assets/js/filters.js', __FILE__ ), array( 'jquery-form', 'pwbe-select2', 'pwbe-context-menu' ), PWBE_VERSION );

            $string_types = array(
                'contains'          => __( 'contains', 'pw-bulk-edit' ),
                'does not contain'  => __( 'does not contain', 'pw-bulk-edit' ),
                'is'                => __( 'is', 'pw-bulk-edit' ),
                'is not'            => __( 'is not', 'pw-bulk-edit' ),
                'begins with'       => __( 'begins with', 'pw-bulk-edit' ),
                'ends with'         => __( 'ends with', 'pw-bulk-edit' ),
            );

            $boolean_types = array(
                'is checked'        => __( 'is checked', 'pw-bulk-edit' ),
                'is not checked'    => __( 'is not checked', 'pw-bulk-edit' ),
            );

            $numeric_types = array(
                'is'                => __( 'is', 'pw-bulk-edit' ),
                'is not'            => __( 'is not', 'pw-bulk-edit' ),
                'is greater than'   => __( 'is greater than', 'pw-bulk-edit' ),
                'is less than'      => __( 'is less than', 'pw-bulk-edit' ),
                'is in the range'   => __( 'is in the range', 'pw-bulk-edit' ),
            );

            $select_types = array(
                'is any of'         => __( 'is any of', 'pw-bulk-edit' ),
                'is none of'        => __( 'is none of', 'pw-bulk-edit' ),
            );

            wp_localize_script( 'pwbe-filters', 'pwbeFilters', array(
                'stringTypes' => array_keys( $string_types ),
                'booleanTypes' => array_keys( $boolean_types ),
                'numericTypes' => array_keys( $numeric_types ),
                'selectTypes' => array_keys( $select_types ),
                'i18n' => array(
                    'stringTypes' => array_values( $string_types ),
                    'booleanTypes' => array_values( $boolean_types ),
                    'numericTypes' => array_values( $numeric_types ),
                    'selectTypes' => array_values( $select_types ),
                )
            ) );

            wp_enqueue_script( 'pwbe-results', plugins_url( '/assets/js/results.js', __FILE__ ), array( 'pwbe-filters' ), PWBE_VERSION );
            wp_localize_script( 'pwbe-results', 'pwbe', array(
                'i18n' => array(
                    'view_name_prompt' => __( 'Name your custom view, for example "My View"', 'pw-bulk-edit' ),
                    'overwrite_view_prompt' => __( 'A view with this name already exists. Do you want to overwrite it?', 'pw-bulk-edit' ),
                ),
                'saveBatchSize' => PWBE_SAVE_BATCH_SIZE,
                'nonces' => array(
                    'save_products' => wp_create_nonce( 'pw-bulk-edit-save-products' ),
                ),
            ) );

            wp_enqueue_script( 'jquery-form' );
            wp_enqueue_script( 'jquery-ui-draggable' );
            wp_enqueue_script( 'jquery-ui-position' );
        }

        wp_register_style( 'pw-bulk-edit-icon', plugins_url( '/assets/css/icon-style.css', __FILE__ ), array(), PWBE_VERSION );
        wp_enqueue_style( 'pw-bulk-edit-icon' );
    }

    public static function starts_with( $needle, $haystack ) {
        $length = strlen( $needle );
        return ( substr( $haystack, 0, $length ) === $needle );
    }

    /**
     * Source: http://wordpress.stackexchange.com/questions/14652/how-to-show-a-hierarchical-terms-list
     * Recursively sort an array of taxonomy terms hierarchically. Child categories will be
     * placed under a 'children' member of their parent term.
     * @param Array   $cats     taxonomy term objects to sort
     * @param Array   $into     result array to put them in
     * @param integer $parentId the current parent ID to put them in
     */
    function sort_terms_hierarchically( array &$cats, array &$into, $parentId = 0 ) {
        foreach ( $cats as $i => $cat ) {
            if ( $cat->parent == $parentId ) {
                $into[$cat->term_id] = $cat;
                unset( $cats[$i] );
            }
        }

        foreach ( $into as $topCat ) {
            $topCat->children = array();
            $this->sort_terms_hierarchically( $cats, $topCat->children, $topCat->term_id );
        }
    }

    function hierarchical_select($categories, $level = 0, $parent = NULL, $prefix = '') {
        $output = '';

        foreach ( $categories as $category ) {
            $output .= "<option value='{$category->slug}'>$prefix {$category->name}</option>\n";

            if ( $category->parent == $parent ) {
                $level = 0;
            }

            if ( count( $category->children ) > 0 ) {
                $output .= $this->hierarchical_select( $category->children, ( $level + 1 ), $category->parent, "$prefix {$category->name} &#8594;" );
            }
        }

        return $output;
    }
}

new PW_Bulk_Edit();

endif;
